presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionProcess.java

```diff
@@ -0,0 +1,265 @@
+        this.nativeExecutionTask = new NativeExecutionTask(session, planFragment, sources, this.workerClient, tableWriteInfo, executor, updateScheduledExecutor, taskManagerConfig);
+        this.errorRetryScheduledExecutor = requireNonNull(errorRetryScheduledExecutor, "errorRetryScheduledExecutor is null");
+        this.errorTracker = nativeExecutionRequestErrorTracker("NativeExecution", uri, maxErrorDuration, errorRetryScheduledExecutor, "getting native process status");
+    }
+
+    /**
+     * Starts the external native execution process. The method will be blocked by connecting to the native process's /v1/info endpoint with backoff retries until timeout.
+     */
+    public void start()
+            throws ExecutionException, InterruptedException, IOException
+    {
+        String executablePath = getProcessWorkingPath(SystemSessionProperties.getNativeExecutionExecutablePath(session));
+        String configPath = Paths.get(executablePath, String.valueOf(port)).toAbsolutePath().toString();
+
+        populateConfigurationFiles(configPath);
+        ProcessBuilder processBuilder = new ProcessBuilder(executablePath, "--v", "1", "--etc_dir", configPath);
+        processBuilder.redirectOutput(ProcessBuilder.Redirect.INHERIT);
+        processBuilder.redirectError(ProcessBuilder.Redirect.INHERIT);
+        try {
+            log.info("Launching %s \nConfig path: %s\n", executablePath, configPath);
+            process = processBuilder.start();
+        }
+        catch (IOException e) {
+            throw new PrestoException(NATIVE_EXECUTION_PROCESS_LAUNCH_ERROR, format("Cannot start %s", processBuilder.command()), e);
+        }
+
+        getServerInfoWithRetry().get();
+    }
+
+    public NativeExecutionTask getTask()
+    {
+        return nativeExecutionTask;
+    }
+
+    @VisibleForTesting
+    public SettableFuture<ServerInfo> getServerInfoWithRetry()
+    {
+        SettableFuture<ServerInfo> future = SettableFuture.create();
+        doGetServerInfo(future);
+        return future;
+    }
+
+    public void close()
+    {
+        if (process.isAlive()) {
+            process.destroy();
+        }
+    }
+
+    private URI getBaseUriWithPort(URI baseUri, int port)
+    {
+        return uriBuilderFrom(baseUri)
+                .port(port)
+                .build();
+    }
+
+    private int getAvailableTcpPort()
+            throws IOException
+    {
+        ServerSocket socket = new ServerSocket(0);
+        int port = socket.getLocalPort();
+        socket.close();
+        return port;
+    }
+
+    private String getNativeExecutionCatalogName(Session session)
+    {
+        checkArgument(session.getCatalog().isPresent(), "Catalog isn't set in the session.");
+        return session.getCatalog().get();
+    }
+
+    private void populateConfigurationFiles(String configBasePath)
+            throws IOException
+    {
+        systemConfig.setHttpServerPort(port);
```

Comment:
nit: It is generally an anti pattern to overwrite the config property in the code. 

Can there be multiple workers on the same box, if so there is a chance for all the workers to keep mutating this property.

Reply:
This code will be executed on the worker side, so the mutation is only visible to the current worker IIUC. The reason we have to pick and assign the port per worker is in our prod environment, there is no port isolation among all the containers running on the same host, so we have to pick unique port per worker to avoid port collision. This system config (`NativeExecutionSystemConfig`) will be passed down to the presto_cpp process eventually.